Ical.Net/Serialization/PropertySerializer.cs

```diff
@@ -39,65 +39,83 @@ public PropertySerializer(SerializationContext ctx) : base(ctx) { }
+
+        // Get the list of parameters we'll be serializing
+        var parameterList = GetParameterList(prop, value);
+
+        // The TZID property of an RDATE/EXDATE collection is owned by the PeriodList that contains it. 
+        // It is allowed to have multiple EXDATE or RDATE collections, each with a different TZID.
+        // Using RecurrenceDate and ExceptionDate classes ensures, that all Periods in the
+        // PeriodList have the same TZID and PeriodKind, which allows PeriodList be serialized in one go.
+        // Here, to determine the timezone, we can safely use the first Period's timezone.
+        if (value is PeriodList periodList)
+        {
+            UpdateTzId(parameterList, periodList);
+        }
 
-            var sb = new StringBuilder();
-            sb.Append(prop.Name);
-            if (parameterList.Any())
+        var sb = new StringBuilder();
+        sb.Append(prop.Name);
+        if (parameterList.Count != 0)
+        {
+            // Get a serializer for parameters
+            var parameterSerializer = sf.Build(typeof(CalendarParameter), SerializationContext) as IStringSerializer;
+            if (parameterSerializer != null)
             {
-                // Get a serializer for parameters
-                var parameterSerializer = sf.Build(typeof(CalendarParameter), SerializationContext) as IStringSerializer;
-                if (parameterSerializer != null)
+                sb.Append(';');
+                var first = true;
+                // Serialize each parameter and append to the StringBuilder
+                foreach (var param in parameterList)
                 {
-                    // Serialize each parameter
-                    // Separate parameters with semicolons
-                    sb.Append(";");
-                    sb.Append(string.Join(";", parameterList.Select(param => parameterSerializer.SerializeToString(param))));
+                    if (!first) sb.Append(';');
+
+                    sb.Append(parameterSerializer.SerializeToString(param));
+                    first = false;
                 }
             }
-            sb.Append(":");
-            sb.Append(value);
-
-            result.Append(TextUtil.FoldLines(sb.ToString()));
         }
+        sb.Append(':');
+        sb.Append(serializedValue);
 
-        // Pop the object off the serialization context.
-        SerializationContext.Pop();
-        return result.ToString();
+        result.FoldLines(sb.ToString());
+    }
+
+    private static IParameterCollection GetParameterList(ICalendarProperty prop, object value)
+        => value is ICalendarDataType dataType ? dataType.Parameters : prop.Parameters;
+
+    private static void UpdateTzId(IParameterCollection parameterList, PeriodList periodList)
+    {
+        if (periodList[0].TzId != null && periodList[0].TzId != "UTC" &&
+            parameterList.All(p => string.Equals("TZID", p.Value, StringComparison.OrdinalIgnoreCase)))
```

Comment:
Shouldn't htis be `.Any()`? But even then, why would we set the TZID only if it already exists? Shouldn't we set it in any case except if floating or UTC?